### ISSUE TITLE
command line engine args creates extra (bad) args

### ISSUE THREAD

**Issue by aowen87** [aowen87/BAR#1824]
There is a long standing bug if one sends -np on the command line for engine it gets parsed but then adds a bunch of other engine args including defaulting to mpirun when it should use what is in the session file.

BAD:  -engineargs -l;mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost  <<< SHOULD NOT BE mpirun but qsub/mpirun

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -l;mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost

GOOD: -engineargs -b;smithp-ash-cs

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -b;smithp-ash-cs

If I add a -l qsub/mpirun when the script works 

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -l;qsub/mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost

-----------------------REDMINE MIGRATION-----------------------
This ticket was migrated from Redmine. As such, not all 
information was able to be captured in the transition. Below is
a complete record of the original redmine ticket.

Ticket number: 2024
Status: Pending
Project: VisIt
Tracker: Bug
Priority: Normal
Subject: command line engine args creates extra (bad) args
Assigned to: 
Category: 
Target version: 3.0
Author: Allen Sanderson
Start: 10/20/2014
Due date: 
% Done: 0
Estimated time: 
Created: 10/20/2014 01:53 pm
Updated: 06/30/2015 07:15 pm
Likelihood: 3 - Occasional
Severity: 3 - Major Irritation
Found in version: 2.7.3
Impact: 
Expected Use: 
OS: All
Support Group: Any
Description:
 There is a long standing bug if one sends -np on the command line for engine it gets parsed but then adds a bunch of other engine args including defaulting to mpirun when it should use what is in the session file.

BAD:  -engineargs -l;mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost  <<< SHOULD NOT BE mpirun but qsub/mpirun

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -l;mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost

GOOD: -engineargs -b;smithp-ash-cs

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -b;smithp-ash-cs

If I add a -l qsub/mpirun when the script works 

Running: cli2.8.1 -start 0 -end 2 -frame 0 -geometry 1024x1280 -format tiff -output movie -sessionfile ../BSFO2.session -nowin -s /uufs/ash.peaks/sys/pkg/VisIt/Install/gcc/2.8.1/linux-x86_64/bin/makemoviemain.py -engineargs -l;qsub/mpirun;-np;64;-nn;8;-b;smithp-ash-cs -launchengine localhost

Comments:
This is a bit more complicated issue than first glance would suggest.The use case is an important one we should try to support. . . Users run VisIt interactively on (small) resources to get plots, views, etc. setup. Then, they save session with the expectation that they can later launch VisIt on different (large) resources, perhaps even multiple instances of VisIt for different windows of the time anmation (Allen has a script to automate this in fact) using command-line arguments to override what is in the session file. If you look at the help messages from visit -fullhelp, it kinda sorta suggests this kind of thing is possible. And, it seems like a common use case we should really support.Now, I thought the GUI's movie-making wizard has allowances for this kind of thing already. Maybe we should just enhance that wizard to spit out a script or set of commands the user can launch that breaks the time-animation up into some number of windows and can launch VisIt in parallel/batch multiple times and then assemble the results into a movie. I will add a separate ticket for this.There appears to be a confluence of information from internallauncher, host profiles and a session file that currently wind up inhibiting a user's attempt to override parallel launch options with command-line options. This is the basic problem. Any parallel launch arguments on the command-line should override any gleaned from the other sources (internallauncher, host profile, session file).For a long time, we've operated with the logic that the presence of a _-np_ and/or _-nn_ option on  the command-line can, alone, serve to launch a parallel VisIt in simple, vanilla, mpirun-compliant settings. This is the current (intentional) design. Do we want to continue supporting that? Perhaps if the user also specifies a session file on the command line, the _-np_ should not imply mpirun for the launch and it should instead continue using the normal logic pathways to form parallel launch arguments.
Let me add a bit more. I was making the assumption that the internal launcher at some point was reading the session file and pulling the launch args from it, then over riding the any args specified on the command line (that is implied from --fullhelp message). This is not true. However, the internal launcher will add args when it is run in parallel. As Mark noted these args assume a simple parallel run. Those args wrongly override what is in the session file.In my movie script I read the session file, gather launch args and then pass them on to VisIt. I needed to do this for creating movies concurrently as the args were needed for qsub. My fix was to do the same for serial mode regardless if there are args on the command line. That solves my issue but only because I have a script.For fun I have added my movie script to this issue. It is was written for a particular machine but could be further generalized.
I removed myself from the assigned field because what I originally volunteered for was to have a conversation with Allen to try to clarify the issues.
